tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -482,18 +482,86 @@ export class WebGPUBackend extends KernelBackend {
     }
   }
 
+  private updateUniforms(
+      uniforms: GPUBindingResource[], uniformDataLengths: number[],
+      uniformData: Uint32Array|Int32Array|Float32Array) {
+    uniformDataLengths.push(uniformData.byteLength);
+    uniforms.push(this.makeUniforms(uniformData));
+  }
+
   public runWebGPUProgram(
       program: webgpu_program.WebGPUProgram, inputs: TensorInfo[],
       outputDtype: DataType,
-      programUniforms?: Uint32Array|Int32Array|Float32Array): TensorInfo {
+      programUniformsData?: Uint32Array|Int32Array|Float32Array): TensorInfo {
     const output = this.makeTensorInfo(program.outputShape, outputDtype);
 
-    let uniformDataLength;
-    let uniforms: GPUBindingResource;
-    if (program.uniforms) {
-      // TODO: handle padding of program-specific uniforms
-      uniformDataLength = programUniforms.byteLength;
-      uniforms = this.makeUniforms(programUniforms);
+    let dimUniformsData: number[] = [];
+    const bufferShapes = inputs.concat(output).map(d => d.shape);
+    if (program.needsShapesUniforms) {
+      let currentOffset = 0;
+      bufferShapes.forEach((d, i) => {
+        // Uniforms.
+        if (d.length === 0) {
+          d = [1];
+        }
+        // Complete std140 layout rules are documented here:
+        // tslint:disable-next-line:max-line-length
+        // https://www.khronos.org/registry/OpenGL/specs/gl/glspec45.core.pdf#page=159
+        let baseAlignment: number;
+        switch (d.length) {
+          case 0:
+            baseAlignment = 1;
+            break;
+          case 1:
+            baseAlignment = 1;
+            break;
+          case 2:
+            baseAlignment = 2;
+            break;
+          case 3:
+            baseAlignment = 4;
+            break;
+          case 4:
+            baseAlignment = 4;
+            break;
+          default:
+            util.assert(false, () => `Unsupported ${d.length}D shape`);
+        }
+
+        const padding =
+            Math.ceil(currentOffset / baseAlignment) * baseAlignment -
+            currentOffset;
+        for (let p = 0; p < padding; ++p) {
+          dimUniformsData.push(0);
+        }
+        dimUniformsData.push(...d);
+        currentOffset += d.length + padding;
+      });
+    }
+
+    const uniformDataLengths: number[] = [];
+    const uniforms: GPUBindingResource[] = [];
+
+    let hasIntProgramUniforms = false;
+    let hasFloatProgramUniforms = false;
+    if (programUniformsData instanceof Int32Array) {
```

Comment:
I don't like merging shapes uniform and program uniform together. It will make the code more complicated and not easy to read. Just separate them into two uniform blocks and you don't need to care about data type Int32Arry or Float32Array.